### ISSUE TITLE
Tests wrapping `main` to be run in a separate process

### ISSUE THREAD

**Issue by mikebentley15** [PRUNERS/FLiT#265]
## Feature Request

**Describe the new feature:**
We have seen a pattern of making tests out of the `main` function of applications or of integration examples or tests.  Currently, wrapping main in a test and calling it more than once violates the fundamental assumptions of writing a `main` function - i.e., that it is called once and only once.  This can potentially lead to runtime nondeterminism (not seen when running the application outside of the test), or even undefined behavior.

Before talking about the proposal, let me give a brief description of how this has been accomplished in the past.  We would follow a pattern like the following

```c++
#define main main_impl
#include "main.cpp"
#undef main

#include <flit.h>

#include <string>

template <typename T>
class MyTest : public flit::TestBase<T> {
public:
    ...
protected:
    virtual flit::Variant run_impl(const std::vector<T> &ti) override {
        // redirect standard out to a char buffer
        int argc = ...; // some value
        char* argv[] = {...}; // some command-line arguments
        int retval = main_impl(argc, argv);
        // convert the standard output into a std::string
        std::string buffer = ...;
        return buffer;
    }
};
...
```

Above, the `main` function is renamed to `main_impl` by using a `#define` before the `#include`, followed by a `#undef`.  Using this approach, we can call the (now renamed) main function inside of the test.

**Suggested change:**
The proposal is to separate out the call to `main` in a separate process.  This would be done by recursively calling the same executable (i.e., `argv[0]`), but in a different mode that only runs the one test one time and outputs the test information to a file.  The parent process then reads the file to get the test results.  This approach allows us to create as clean of an environment as we can conceivably create for the user's `main` function, and it will only be called once within a process.  We can still do timing because we can spawn a child process for each time we want to execute the test.

But, we do not want to take this approach for all FLiT tests.  For unit tests, creating separate processes is too much unnecessary overhead.  To this end, we believe we could have another test base class that would spawn another process, something like

```c++
template <typename T>
class SeparateProcessTestBase : public TestBase<T> {...};
```

This class would have the same pure virtual functions but would reimplement `run` to either spawn a separate process or to call `run_impl` and output to a file.  Whether that is in a separate function or not remains as an implementation detail.  It would be great if we could reuse `flit::TestBase::run()` from `flit::SeparateProcessTestBase::run()`.

Below is a workout Ian and I did on the whiteboard to handle this problem and also the problem of MPI, specifically addressing how we can have the test call `MPI_Init()` and `MPI_Finalize()`.  The MPI aspect is a separate issue that needs to be implemented at the same time this one is done, since doing one without the other breaks functionality.

![flit-mpi-whiteboard-plan](https://user-images.githubusercontent.com/8606516/57964015-f9caab00-78ea-11e9-934c-dcee54157920.jpg)

**Alternative approaches:**
A complete rearchitecting of the testing framework could be done instead.  We could have a test runner either in C++ or in Python.  This kind of approach would probably have ALL tests be done as separate processes and everything is managed instead by a test runner.  We would probably still want all of the tests compiled into a single executable just so that disk usage is not unnecessarily consumed (think about 8 tests with 250 compilations, each one being 5 MB, that would be 10 GB of space - not cheap).

**Comment by mikebentley15** [comment]
This hasn't yet been merged in
